File: prefix-sum/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prefix = [1]*len(nums)
        postfix = [1]*len(nums)

        for i in range(1, len(nums)):
            prefix[i] = prefix[i-1] * nums[i-1]

        for i in range(len(nums)-2,-1, -1):
            postfix[i] = nums[i+1] * postfix[i+1]
        print(postfix)
        for i in range(len(nums)):
            nums[i] = prefix[i] * postfix[i]
        print(prefix)
        
        return nums  

File: array/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] prefix = new int[nums.length];
        Arrays.fill(prefix, 1);

        int[] post = new int[nums.length];
        Arrays.fill(post, 1);

        for (int i=1; i < nums.length; i++){
            prefix[i] = nums[i-1] * prefix[i-1]; 
        }

        for (int i = nums.length-2; i >=0; i--){
            post[i] = nums[i+1] * post[i+1];
        }

        int[] store = new int[nums.length];
        for (int i=0; i < nums.length; i++){
            store[i] = prefix[i] * post[i];
        }
        return store;

    }
}